CodeIndigo/Indigo/Doc/man/man3/Indigo.3
.TH "Indigo" 3 "Mon May 5 2014" "Version 200" "Indigo Engine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Indigo \- 
.SH SYNOPSIS
.br
.PP
.SS "Functions"

.in +1c
.ti -1c
.RI "void \fBInitialize\fP (const char *window_name, glm::vec3 background, const int window_width, const int window_height, const int max_framerate, const bool fullscreen)"
.br
.ti -1c
.RI "int \fBRun\fP (void)"
.br
.ti -1c
.RI "void \fBClose\fP (void)"
.br
.ti -1c
.RI "void \fBReshape\fP (GLFWwindow *window, int width, int height)"
.br
.ti -1c
.RI "void \fBKey_Action\fP (GLFWwindow *window, int key, int scancode, int action, int modifiers)"
.br
.ti -1c
.RI "bool \fBPressed\fP (unsigned char key)"
.br
.ti -1c
.RI "void \fBMouse_Button\fP (GLFWwindow *window, int button, int state, int modifiers)"
.br
.ti -1c
.RI "void \fBMouse_Moved\fP (GLFWwindow *window, double x, double y)"
.br
.ti -1c
.RI "void \fBFPS_Mouse_Callback\fP (int x, int y, \fBObject\fP *player, float sensitivity)"
.br
.ti -1c
.RI "void \fBFPS_Mouse_Function\fP (int x, int y)"
.br
.ti -1c
.RI "void \fBFPS_Mouse\fP (\fBObject\fP *player, float sensitivity)"
.br
.ti -1c
.RI "void \fBError_Found\fP (int type, const char *message)"
.br
.ti -1c
.RI "void \fBUpdate\fP (const float time)"
.br
.ti -1c
.RI "void \fBRender\fP (void)"
.br
.ti -1c
.RI "float \fBElapsed\fP ()"
.br
.ti -1c
.RI "float \fBFast_Float\fP (const char *stream, int *output, const int start)"
.br
.ti -1c
.RI "char \fBGet_Shifted_Character\fP (const char key)"
.br
.ti -1c
.RI "void \fBError_Dump\fP (void)"
.br
.ti -1c
.RI "glm::vec3 \fBWhite_Color\fP (1\&.0, 1\&.0, 1\&.0)"
.br
.ti -1c
.RI "glm::vec3 \fBBlack_Color\fP (0\&.0, 0\&.0, 0\&.0)"
.br
.ti -1c
.RI "glm::vec3 \fBSky_Color\fP (0\&.5, 0\&.8, 1\&.0)"
.br
.ti -1c
.RI "glm::vec3 \fBRed_Color\fP (0\&.4, 0\&.0, 0\&.0)"
.br
.ti -1c
.RI "glm::vec3 \fBGreen_Color\fP (0\&.0, 0\&.8, 0\&.0)"
.br
.ti -1c
.RI "glm::vec3 \fBBlue_Color\fP (0\&.0, 0\&.0, 0\&.6)"
.br
.ti -1c
.RI "glm::vec3 \fBLight_Blue_Color\fP (0\&.4, 0\&.7, 0\&.9)"
.br
.ti -1c
.RI "glm::vec3 \fBOrange_Color\fP (1\&.0, 0\&.4, 0\&.0)"
.br
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "\fBWorld\fP \fBCurrent_World\fP"
.br
.ti -1c
.RI "GLFWwindow * \fBWindow\fP = nullptr"
.br
.ti -1c
.RI "void(* \fBKey_Pressed_Function\fP )(int key)"
.br
.ti -1c
.RI "void(* \fBKey_Released_Function\fP )(int key)"
.br
.ti -1c
.RI "void(* \fBMouse_Button_Function\fP )(int button, int state, float x, float y)"
.br
.ti -1c
.RI "void(* \fBMouse_Moved_Function\fP )(int x, int y)"
.br
.ti -1c
.RI "void(* \fBRelative_Mouse_Moved_Function\fP )(int x, int y)"
.br
.ti -1c
.RI "void(* \fBUpdate_Function\fP )(float time)"
.br
.ti -1c
.RI "void(* \fBRender_Function\fP )(void)"
.br
.ti -1c
.RI "void(* \fBError_Function\fP )(int type, const char *message)"
.br
.ti -1c
.RI "bool \fBKeys\fP [512]"
.br
.ti -1c
.RI "bool \fBKeys_Pressed\fP [512]"
.br
.ti -1c
.RI "bool \fBShift\fP"
.br
.ti -1c
.RI "bool \fBControl\fP"
.br
.ti -1c
.RI "bool \fBAlt\fP"
.br
.ti -1c
.RI "bool \fBLeft_Mouse\fP = false"
.br
.ti -1c
.RI "bool \fBRight_Mouse\fP = false"
.br
.ti -1c
.RI "bool \fBMiddle_Mouse\fP = false"
.br
.ti -1c
.RI "glm::vec2 \fBMouse_Position\fP = glm::vec2(0, 0)"
.br
.ti -1c
.RI "glm::vec2 \fBMouse_Position_Screen\fP = glm::vec2(0, 0)"
.br
.ti -1c
.RI "unsigned int \fBVAO\fP"
.br
.ti -1c
.RI "int \fBScreen_Width\fP = 0"
.br
.ti -1c
.RI "int \fBScreen_Height\fP = 0"
.br
.ti -1c
.RI "float \fBAspect_Ratio\fP = 1"
.br
.ti -1c
.RI "float \fBFrame_Length_Minimum\fP = 0"
.br
.ti -1c
.RI "int \fBActual_FPS\fP = 60"
.br
.ti -1c
.RI "glm::vec3 \fBWhite_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBBlack_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBSky_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBRed_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBGreen_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBBlue_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBLight_Blue_Color\fP"
.br
.ti -1c
.RI "glm::vec3 \fBOrange_Color\fP"
.br
.in -1c
.SH "Function Documentation"
.PP 
.SS "glm::vec3 Indigo::Black_Color (0\&.0, 0\&.0, 0\&.0)"

.SS "glm::vec3 Indigo::Blue_Color (0\&.0, 0\&.0, 0\&.6)"

.SS "void Indigo::Close (void)"

.SS "float Indigo::Elapsed ()\fC [inline]\fP"

.SS "void Indigo::Error_Dump (void)"

.SS "void Indigo::Error_Found (inttype, const char *message)"

.SS "float Indigo::Fast_Float (const char *stream, int *output, const intstart)"

.SS "void Indigo::FPS_Mouse (\fBObject\fP *player, floatsensitivity)"

.SS "void Indigo::FPS_Mouse_Callback (intx, inty, \fBObject\fP *player, floatsensitivity)"

.SS "void Indigo::FPS_Mouse_Function (intx, inty)"

.SS "char Indigo::Get_Shifted_Character (const charkey)"

.SS "glm::vec3 Indigo::Green_Color (0\&.0, 0\&.8, 0\&.0)"

.SS "void Indigo::Initialize (const char *window_name, glm::vec3background, const intwindow_width, const intwindow_height, const intmax_framerate, const boolfullscreen)"

.SS "void Indigo::Key_Action (GLFWwindow *window, intkey, intscancode, intaction, intmodifiers)"

.SS "glm::vec3 Indigo::Light_Blue_Color (0\&.4, 0\&.7, 0\&.9)"

.SS "void Indigo::Mouse_Button (GLFWwindow *window, intbutton, intstate, intmodifiers)"

.SS "void Indigo::Mouse_Moved (GLFWwindow *window, doublex, doubley)"

.SS "glm::vec3 Indigo::Orange_Color (1\&.0, 0\&.4, 0\&.0)"

.SS "bool Indigo::Pressed (unsigned charkey)"

.SS "glm::vec3 Indigo::Red_Color (0\&.4, 0\&.0, 0\&.0)"

.SS "void Indigo::Render (void)"

.SS "void Indigo::Reshape (GLFWwindow *window, intwidth, intheight)"

.SS "int Indigo::Run (void)"

.SS "glm::vec3 Indigo::Sky_Color (0\&.5, 0\&.8, 1\&.0)"

.SS "void Indigo::Update (const floattime)"

.SS "glm::vec3 Indigo::White_Color (1\&.0, 1\&.0, 1\&.0)"

.SH "Variable Documentation"
.PP 
.SS "int Indigo::Actual_FPS = 60"

.SS "bool Indigo::Alt"

.SS "float Indigo::Aspect_Ratio = 1"

.SS "glm::vec3 Indigo::Black_Color"

.SS "glm::vec3 Indigo::Blue_Color"

.SS "bool Indigo::Control"

.SS "\fBWorld\fP Indigo::Current_World"

.SS "void(* Indigo::Error_Function)(int type, const char *message)"

.SS "float Indigo::Frame_Length_Minimum = 0"

.SS "glm::vec3 Indigo::Green_Color"

.SS "void(* Indigo::Key_Pressed_Function)(int key)"

.SS "void(* Indigo::Key_Released_Function)(int key)"

.SS "bool Indigo::Keys"

.SS "bool Indigo::Keys_Pressed"

.SS "bool Indigo::Left_Mouse = false"

.SS "glm::vec3 Indigo::Light_Blue_Color"

.SS "bool Indigo::Middle_Mouse = false"

.SS "void(* Indigo::Mouse_Button_Function)(int button, int state, float x, float y)"

.SS "void(* Indigo::Mouse_Moved_Function)(int x, int y)"

.SS "glm::vec2 Indigo::Mouse_Position = glm::vec2(0, 0)"

.SS "glm::vec2 Indigo::Mouse_Position_Screen = glm::vec2(0, 0)"

.SS "glm::vec3 Indigo::Orange_Color"

.SS "glm::vec3 Indigo::Red_Color"

.SS "void(* Indigo::Relative_Mouse_Moved_Function)(int x, int y)"

.SS "void(* Indigo::Render_Function)(void)"

.SS "bool Indigo::Right_Mouse = false"

.SS "int Indigo::Screen_Height = 0"

.SS "int Indigo::Screen_Width = 0"

.SS "bool Indigo::Shift"

.SS "glm::vec3 Indigo::Sky_Color"

.SS "void(* Indigo::Update_Function)(float time)"

.SS "unsigned int Indigo::VAO"

.SS "glm::vec3 Indigo::White_Color"

.SS "GLFWwindow * Indigo::Window = nullptr"

.SH "Author"
.PP 
Generated automatically by Doxygen for Indigo Engine from the source code\&.


CodeIndigo/Indigo/Indigo.cpp
// Basic OpenGL functions, also takes care of most of the background work.
// These function will mostly be called in Main.

#include "GL/glew.h"

#include "Indigo.h"


namespace Indigo
{
	// Initializes window and rendering matrices. Really ought to be called first.
	void Initialize(const char * window_name, glm::vec3 background, const int window_width,
		const int window_height, const int max_framerate, const bool fullscreen)
	{
		// Initiate glut
		glfwSetErrorCallback(Error_Found);
		glfwInit();
		glfwWindowHint(GLFW_SAMPLES, 4);
		glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 2);
		glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 1);
		//glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
		if (fullscreen)
		{
			const GLFWvidmode * monitor = glfwGetVideoMode(glfwGetPrimaryMonitor());
			Window = glfwCreateWindow(window_width ? window_width : monitor->width, window_height ? window_height : monitor->height, window_name, glfwGetPrimaryMonitor(), NULL);
		}
		else
		{
			Window = glfwCreateWindow(window_width, window_height, window_name, NULL, NULL);
		}
		glfwSetFramebufferSizeCallback(Window, Reshape);
		glfwSetCursorPosCallback(Window, Mouse_Moved);
		glfwSetMouseButtonCallback(Window, Mouse_Button);
		glfwSetKeyCallback(Window, Key_Action);
		glfwMakeContextCurrent(Window);

		// GLEW does all this magical driver OGL stuff
		glewExperimental = GL_TRUE;
		if (glewInit())
		{
			std::cout << "There was this GLEW thing, and it was all like, NO!" << std::endl << "And we were sad. Try reinstalling."
				<< std::endl << "For more info contact the creator of your game." << std::endl;
		}

		glClearColor(background[0], background[1], background[2], 1.0);
		if (max_framerate == 0)
		{
			Frame_Length_Minimum = 0;
		}
		else
		{
			Frame_Length_Minimum = 1000.0 / max_framerate;
		}

		//glGenVertexArrays(1, &VAO);
		//glBindVertexArray(VAO);

		glEnable(GL_DEPTH_TEST);
		glDepthFunc(GL_LESS);
		glEnable(GL_DITHER);
		glEnable(GL_BLEND);
		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
		//glEnable(GL_TEXTURE_2D);
		// Setup fog
		//glEnable(GL_FOG);
		//glFogfv(GL_FOG_COLOR, White_Color);
		//glHint(GL_FOG_HINT, GL_DONT_CARE);
		//glFogf(GL_FOG_DENSITY, 0.002);

		// Enable rendering options
		Reshape(Window);

		//Indigo::Error_Dump();

		return;
	}

	// Starts the main loop with update, render, and input
	int Run(void)
	{
		Render();
		glfwSwapBuffers(Indigo::Window);
		glfwPollEvents();
		float last = 0;
		while (!glfwWindowShouldClose(Window))
		{
			float now = Elapsed();
			float delta = now - last;
			if (delta >= Frame_Length_Minimum)
			{
				last = now;
				Update(delta);
				Render();
				glfwSwapBuffers(Indigo::Window);
				glfwPollEvents();
			}
		}
		glfwTerminate();
		return 0;

	}

	void Close(void)
	{
		glfwSetWindowShouldClose(Window, true);
		return;
	}

	// Acts for when the window reshapes
	void Reshape(GLFWwindow * window, int width, int height)
	{
		int window_width;
		int window_height;
		if (0 == width || 0 == height)
		{
			glfwGetFramebufferSize(Window, &window_width, &window_height);
		}
		else
		{
			window_width = width;
			window_height = height;
			glViewport(0, 0, width, height);
		}
		Screen_Width = window_width;
		Screen_Height = window_height;
		Aspect_Ratio = (float) window_width / (float) window_height;
	}

	// Acts for Keys which act once, and stores for multi-acting Keys
	void Key_Action(GLFWwindow * window, int key, int scancode, int action, int modifiers)
	{
		if (key >= 'A' && key <= 'Z')
		{
			key += 'a' - 'A';
		}

		Shift = (modifiers & GLFW_MOD_SHIFT) != 0;
		Control = (modifiers & GLFW_MOD_CONTROL) != 0;
		Alt = (modifiers & GLFW_MOD_ALT) != 0;

		if (action == GLFW_PRESS)
		{
			if (Key_Pressed_Function)
			{
				Key_Pressed_Function(key);
			}
			Keys[key] = true;
			Keys_Pressed[key] = true;
		}
		if (action == GLFW_RELEASE)
		{
			if (Key_Released_Function)
			{
				Key_Released_Function(key);
			}
			Keys[key] = false;
		}
		return;
	}

	// Checks whether a key has just been pressed
	bool Pressed(unsigned char key)
	{
		bool pressed = Keys_Pressed[key];
		Keys_Pressed[key] = false;
		return pressed;
	}

	// Acts for when the mouse is pressed or released
	void Mouse_Button(GLFWwindow * window, int button, int state, int modifiers)
	{
		Shift = (modifiers & GLFW_MOD_SHIFT) != 0;
		Control = (modifiers & GLFW_MOD_CONTROL) != 0;
		Alt = (modifiers & GLFW_MOD_ALT) != 0;
		if (button == GLFW_MOUSE_BUTTON_LEFT)
		{
			Left_Mouse = state == GLFW_PRESS;
		}
		if (button == GLFW_MOUSE_BUTTON_RIGHT)
		{
			Right_Mouse = state == GLFW_PRESS;
		}
		if (button == GLFW_MOUSE_BUTTON_MIDDLE)
		{
			Middle_Mouse = state == GLFW_PRESS;
		}
		if (Mouse_Button_Function)
		{
			Mouse_Button_Function(button, state, Mouse_Position.x, Mouse_Position.y);
		}
		return;
	}

	// Acts for when the mouse is moved
	void Mouse_Moved(GLFWwindow * window, double x, double y)
	{
		Mouse_Position_Screen.x = x;
		Mouse_Position_Screen.y = y;
		Mouse_Position.x = (x - (Screen_Width - Screen_Height) / 2) * 2.0 / Screen_Height - 1;
		Mouse_Position.y = -1 * (y * 2.0 / Screen_Height - 1);
		if (Relative_Mouse_Moved_Function)
		{
			static int last_x = 0;
			static int last_y = 0;
			Relative_Mouse_Moved_Function(x - last_x, y - last_y);
			last_x = x;
			last_y = y;
		}
		if (Mouse_Moved_Function)
		{
			Mouse_Moved_Function(x, y);
		}
		return;
	}

	// Default FPS-style mouse code, used intrinsically. Call FPS_Mouse to set up.
	void FPS_Mouse_Callback(int x, int y, Object * player, float sensitivity)
	{
		// Store the data from this call if it's required so that the callback works
		static float stored_sensitivity;
		static Object * stored_player;
		if (player)
			stored_player = player;
		if (sensitivity)
			stored_sensitivity = sensitivity;
		if (!sensitivity && stored_sensitivity)
			sensitivity = stored_sensitivity;
		if (!player && stored_player)
			player = stored_player;
		if ((!sensitivity && !stored_sensitivity) || (!player && !stored_player))
		{
			std::cout << "Error. FPS_Mouse_Callback is a function used intrinsically by Indigo Engine." << std::endl
				<< "Do not call unless you know what you are doing. Use FPS_Mouse(playerObject) to setup." << std::endl
				<< "Resetting mouse callback to nullptr to prevent further error. Passing as warning." << std::endl;
			Relative_Mouse_Moved_Function = nullptr;
			return;
		}
		float y_angle = player->Facing.Get_Y_Angle() + y * -1 * sensitivity;
		if (!(y_angle > 89 && y_angle < 271))
		{
			player->Facing.Add_Direction(0.0, x * sensitivity, y * -1 * sensitivity);
		}
		else
		{
			if (y_angle < 180)
			{
				player->Facing.Set_Direction(1.0, player->Facing.Get_X_Angle() + x * sensitivity, 89);
			}
			else
			{
				player->Facing.Set_Direction(1.0, player->Facing.Get_X_Angle() + x * sensitivity, 271);
			}
		}
	}

	// Default parameter call, needed for FPS mouse callback. Used intrinsically, do not call.
	void FPS_Mouse_Function(int x, int y)
	{
		FPS_Mouse_Callback(x, y);
	}

	// Default FPS-style mouse for looking around. Set an object pointer that sets onto your camera.
	// Then, use Indigo::Current_World.camera.facing = player.facing;
	void FPS_Mouse(Object * player, float sensitivity)
	{
		static bool enable = true;
		if (enable)
		{
			glfwSetInputMode(Window, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
			FPS_Mouse_Callback(0, 0, player, sensitivity);
			Relative_Mouse_Moved_Function = FPS_Mouse_Function;
		}
		else
		{
			glfwSetInputMode(Window, GLFW_CURSOR, GLFW_CURSOR_NORMAL);
			Relative_Mouse_Moved_Function = nullptr;
		}
		enable = !enable;
		return;
	}

	// Acts for when an error is encountered
	void Error_Found(int type, const char * message)
	{
		std::cout << "Error. " << message << std::endl;
		if (Error_Function)
		{
			Error_Function(type, message);
		}
	}

	// Updates world
	void Update(const float time)
	{
		Actual_FPS = 1000 / time;
		if (Update_Function)
		{
			Update_Function(time);
		}
		Current_World.Update(time);
		return;
	}

	// Renders world
	void Render(void)
	{
		if (Render_Function)
		{
			Render_Function();
		}
		Current_World.Render();
		return;
	}

	// Get elapsed time in the game, optional minus for partial times, in milliseconds
	inline float Elapsed()
	{
		return glfwGetTime() * 1000;
	}

	// Get the floating point equivalent and the length of string with standard notation assuming start of float is at start
	float Fast_Float(const char * stream, int* output, const int start)
	{
		int i = start;
		for (; stream[i] == ' '; ++i);
		int negative = 1;
		if (stream[i] == '-')
		{
			negative = -1;
			++i;
		}
		float result = 0;
		for (; stream[i] >= '0' && stream[i] <= '9'; ++i)
		{
			result = result * 10 + (stream[i] - '0');
		}
		if (stream[i] != '.')
		{
			result *= negative;
			if (output)
			{
				*output = i;
			}
			return result;
		}
		++i;
		float precision = 0.1;
		for (; stream[i] >= '0' && stream[i] <= '9'; ++i)
		{
			result += precision * (stream[i] - '0');
			precision *= 0.1;
		}
		result *= negative;
		if (output)
		{
			*output = i;
		}
		return result;
	}


	// Get the letter as if the key was pressed with shift. Useful for Key Callbacks / checking
	char Get_Shifted_Character(const char key)
	{
		if (key == '`')
			return '~';
		if (key == '1')
			return '!';
		if (key == '2')
			return '@';
		if (key == '3')
			return '#';
		if (key == '4')
			return '$';
		if (key == '5')
			return '%';
		if (key == '6')
			return '^';
		if (key == '7')
			return '&';
		if (key == '8')
			return '^';
		if (key == '9')
			return '(';
		if (key == '0')
			return ')';
		if (key == '-')
			return '_';
		if (key == '=')
			return '+';
		if (key >= 'a' && key <= 'z')
			return key + 'A' - 'a';
		if (key == '[')
			return '{';
		if (key == ']')
			return '}';
		if (key == '\\')
			return '|';
		if (key == ';')
			return ':';
		if (key == '\'')
			return '"';
		if (key == ',')
			return '<';
		if (key == '.')
			return '>';
		if (key == '/')
			return '?';
		return key;
	}


	// Get all the errors since last error dump to main console window
	void Error_Dump(void)
	{
		std::cout << "------ Begin Error Dump --------" << std::endl;
		int code = -1;
		while (code != GL_NO_ERROR)
		{
			code = glGetError();
			if (code != GL_NO_ERROR)
			{
				std::cout << code << std::endl;
			}
		}
		std::cout << "------- End Error Dump ---------" << std::endl;
	}


	// Stores the current world to render
	World Current_World;

	// Stores the window we're rendering onto
	GLFWwindow * Window = nullptr;


	// Stores the function to call when a key is pressed
	void(*Key_Pressed_Function)(int key);

	// ... when a key is released
	void(*Key_Released_Function)(int key);

	// ... when the mouse is pressed or released. Given in 2D_Object space
	void(*Mouse_Button_Function)(int button, int state, float x, float y);

	// ... when the mouse is moved
	void(*Mouse_Moved_Function)(int x, int y);

	// ... when the mouse is moved, given relative to the center.
	// Also hides mouse when defined.
	void(*Relative_Mouse_Moved_Function)(int x, int y);
	
	// ... every time the world updates
	void(*Update_Function)(float time);

	// ... just before the rendering of objects in the world
	void(*Render_Function)(void);

	// ... on an error
	void(*Error_Function)(int type, const char * message);


	// Members with the index of a key which is currently down are true, always lowercase. Abnormals are GLFW_ESCAPE, etc.
	bool Keys[512];

	// Don't use, used internally. Use Pressed('a') instead.
	bool Keys_Pressed[512];

	// Stores whether modifiers are pressed
	bool Shift;

	bool Control;

	bool Alt;

	// Stores whether mouse buttons are down
	bool Left_Mouse = false;

	bool Right_Mouse = false;

	bool Middle_Mouse = false;

	glm::vec2 Mouse_Position = glm::vec2(0, 0);

	glm::vec2 Mouse_Position_Screen = glm::vec2(0, 0);


	// The index for the VAO thingy, not probably important.
	unsigned int VAO;


	// Stores the width of the screen
	int Screen_Width = 0;

	// Stores the height of the screen
	int Screen_Height = 0;

	// Stores the aspect ratio of the screen
	float Aspect_Ratio = 1;

	// Stores the milliseconds to add between each frame
	float Frame_Length_Minimum = 0;

	// Stores the current actual FPS of the update loop
	int Actual_FPS = 60;


	// Colors

	glm::vec3 White_Color(1.0, 1.0, 1.0);

	glm::vec3 Black_Color(0.0, 0.0, 0.0);

	glm::vec3 Sky_Color(0.5, 0.8, 1.0);

	glm::vec3 Red_Color(0.4, 0.0, 0.0);

	glm::vec3 Green_Color(0.0, 0.8, 0.0);

	glm::vec3 Blue_Color(0.0, 0.0, 0.6);

	glm::vec3 Light_Blue_Color(0.4, 0.7, 0.9);

	glm::vec3 Orange_Color(1.0, 0.4, 0.0);

}